### PR TITLE
Generate package typings

### PR DESCRIPTION
**Pull request by chriskrycho**
- Generates `types/<my-app>/index.d.ts` and puts the prototype extensions there if it's a normal app (#88)
- Moves types for `config/environment.js` to `types/<my-app>/config/environment.d.ts`
- Adds mapping to look up things in `types`

@dwickern @dfreeman all of this look sane?

### CODE REVIEW

File: blueprints/ember-cli-typescript/index.js [hunk 1]
```diff
@@ -45,8 +54,24 @@ module.exports = {
...
+        return useAppDeclarations ? APP_DECLARATIONS : '';
+      },
+    };
+  },
+
+  fileMapTokens(/*options*/) {
+    // Return custom tokens to be replaced in your files.
+    return {
+      __app_name__(options) {
+        return options.dasherizedModuleName;
```
Comment (dfreeman) [inline blueprints/ember-cli-typescript/index.js]: We want this to be `'dummy'` for addons, right? I think `dasherizedModuleName` will be the addon name here in that situation.

Comment (chriskrycho) [reply blueprints/ember-cli-typescript/index.js]: This was the other bit I was wondering about, yeah.

Comment (chriskrycho) [reply blueprints/ember-cli-typescript/index.js]: Updated to that effect, @dfreeman.

**Comment by chriskrycho** [comment]
Concur: I think having them present and documenting it makes it *easy* for people to turn on if they need them, but makes the default do the "right thing" as it were. I'll merge as soon as AppVeyor goes green. Which will be a bit longer now that I've merged in `master` again, of course. 😭

**Pull request merged**